WIP: .hdbcds to .hdbtable migration guide (#825)
cap-js/docs: cds/compiler-hdbcds-to-hdbtable.md

---
shorty: Moving from .hdbcds to .hdbtable

# layout: cds-ref

> redirect_from: releases/compiler-v2

Reviewer: This redirect is not needed, right? It also doesn't work:
![image](https://github.com/user-attachments/assets/5446f38c-6759-44b6-ae01-6d81b40f753a)

Reviewer 2: Yes. I guess that was a copy and past thing in the initial version of this guide. Just remove it.

Reviewer: Done

> #status: internal

Reviewer: I guess it should be release, right? ;)
```suggestion
status: released
```

::: info Deployment Format
The format only determines the "medium" through which your database model is brought to the database. The resulting database tables and views are still the same, independent of the format.
:::

## Migration Procedure

If your database deployment currently uses `hdbcds`, it's recommended to switch to the default format, `hdbtable`. This guide assumes you use cds7 or higher. Make sure you read the entire guide before starting the migration process.

1. Ensure that your current data model matches the deployed data model.

2. Switch the deployment format from `hdbcds` to the default `hdbtable`. You can do this by removing option `cds.requires.db.deploy-format` from your configuration files.
   <!-- requires @sap/cds v7 -->
   <!-- this option is not documented, but mentioned in release notes and the changelog -->

3. Undeploy the CAP-generated _.hdbcds_ files by adding an entry to `db/undeploy.json`:
    ::: code-group
    ```json [db/undeploy.json]
    [
      ...,
      "src/gen/**/*.hdbcds"
    ]
    ```
    :::
    
    :::details Possible error message
    Without this entry, during HDI deployment you get errors like the following:

>     ```

Reviewer: ```suggestion
    ```log
```

existing data into this shadow table.

If the table doesn't contain much data, that should not make that much of a difference.
With huge amounts of data, you have to expect a longer and more resource intensive deployment.

:::

## @sql.append and @sql.prepend

Annotations [`@sql.append/prepend`](../guides/databases#sql-prepend-append) allow to
add native SQL clauses to the generated .hdbtable files,
or to add "native" HANA CDS clauses to the generated .hdbcds files, respectively.

If you have used these annotations in your model, a simple switchover from hdbcds to hdbtable
very likely is not possible, as such an annotation written for hdbcds in general is not valid
for hdbtable. You have to adapt your model prior to the migration.

As we don't know what clauses you have used, we cannot offer any further guidance here.

## `WITH ASSOCIATIONS`

Associations cause issues in the .hdbcds to .hdbtable handover.
For each entity that has associations, the resulting table or view contains a `WITH ASSOCIATIONS` clause,
representing native HANA associations.
This clause slightly differs, depending on whether it originates from a .hdbcds or a .hdbtable/.hdbview file,

> which would causes a full table migration.

Reviewer 3: For me this implied that `HDBCDS` files also have `WITH ASSOCIATIONS`.

They come with "Associations" but as part of the column definitions.

Reviewer 2: hi @patricebender , I have rephrased the section on associations. Is it clearer now?

> The CAP Java runtime and the CAP Nodejs runtime with the new SAP HANA service (`@cap-js/hana`, default in CDS 8)
> don't need the `WITH ASSOCIATIONS` clause any more. The workaround is to remove the `WITH ASSOCIATIONS` clause
> __before__ the actual hdbcds to hdbtable migration.

Reviewer 3: ```suggestion
The CAP Java runtime and the CAP Nodejs runtime with the new SAP HANA service (`@cap-js/hana`, default in CDS 8)
don't need the `WITH ASSOCIATIONS` clause any more. The workaround is to remove the `associations` from the `hdbcds` sources
__before__ the actual `hdbcds` to `hdbtable` migration.
```

> First switch of the generation of the `WITH ASSOCIATIONS` clause (both for hdbcds and hdbtable):

Reviewer 3: ```suggestion
First switch of the generation of the `associations` (that option accounts for "associations" in the `hdbcds` sources as well as for the `WITH ASSOCIATIONS` found in the `hdbtable` sources):
```

::: code-group

```json [cdsrc.json]
{
  "sql": {
    "native_hana_associations": false
  }
}
```
:::
<!-- this option is available only with CDS 8 -->

> Then run a new build and deploy the newly generated `.hdbcds` files.
> The resulting database tables and views shouldn't contain the `WITH ASSOCIATIONS` clause anymore.
> In contrast to the hdbtable plugin, the hdbcds plugin is able to handle removal of the
> native associations without a full table migration.

Reviewer 3: ```suggestion
Then run a new build and deploy the newly generated `.hdbcds` files.
The resulting database tables and views shouldn't contain any `associations` anymore.
In contrast to the hdbtable plugin, the hdbcds plugin is able to handle removal of the
native associations without a full table migration.
```

:::

<!--
  full syntax in mta for try_fast is:
  com.sap.hana.di.table/try_fast_table_migration: "true"
-->

## Multiline Doc Comments

This is only relevant if you have switched on [Doc Comments](../cds/cdl#doc-comments-%E2%80%94)
and if you have enabled translation of doc comments to the `COMMENT` feature in the database.

<!--
::: code-group
```json [cdsrc.json]
{
  "docs": true,
  "hana": {
    "comments": true
  }
}
```
:::
-->

Doc comments can span across multiple lines:

```cds
entity Employees {
  key ID : Integer;
  /**
    * I am the description for "name".
    * I span across multiple lines.
    */
  name : String;
}
```

An entity with such a multi-line doc comment can't be seamlessly migrated, because
doc comments are passed to the database differently for hdbcds and hdbtable, respectively.
This would result in a full table migration when changing from hdbcds to hdbtable.

If you don't actually need the comments in the database, you can remove them as a preparation step
__before__ you do the hdbcds to hdbtable migration.
This is similar to the workaround described above for the `WITH ASSOCIATIONS` clause.

> First disable the doc comments by setting option `hana.comments` to `false`.

Reviewer 2: done

Then run a new build and deploy the newly generated `.hdbcds` files.
The resulting database tables and views shouldn't contain the `COMMENT` clause anymore.
In contrast to the hdbtable plugin, the hdbcds plugin is able to handle removal of the

> `COMMENT`s without a full table migration.

Reviewer 2: @patricebender  have changed it, please check

> `COMMENT`s without a full table migration.

Reviewer 3: bueno

If you are already using SAP HANA Cloud, there is no SAP HANA CDS.
:::

The deploy format `hdbcds` for SAP HANA together with the function [`to.hdbcds`](../node.js/cds-compile#hdbcds) have been deprecated with `@sap/cds-compiler@5` and `@sap/cds@8`. Users are advised to switch to the default format `hdbtable`. This guide provides step-by-step instructions for making the switch, including potential issues and workarounds, such as handling annotations `@sql.prepend/append` and dealing with associations.

New CDS features will not be available for deploy format `hdbcds`, and it is going to be removed with one of the
next major releases.

::: info Deploy Format
The deploy format determines only the "medium" how your database model is brought to the database.
The resulting database tables and views are the same, independent of the deploy format.
:::

## Migration Procedure

If your database deployment currently uses `hdbcds`, it's recommended to switch to the default format, `hdbtable`. This guide assumes you use cds7 or higher. Make sure to read the entire guide before starting the migration process.

1. Ensure your current data model matches the deployed data model.

>    <!-- **TBD** must it be exactly the same? Does it also work if the current model

Author: I cannot answer this with confidence, no. I think we should have "Ensure your current data model matches the deployed data model." as a hard requirement!

>    <!-- **TBD** must it be exactly the same? Does it also work if the current model

Reviewer 2: Ok, then we keep the text as is, I will remove the TBD comment

                is changed in comparison to the last deployed model. -->

2. Switch the deploy format from `hdbcds` to the default `hdbtable`. You can do this by removing option `cds.requires.db.deploy-format` from your configuration file(s).
   <!-- requires @sap/cds v7 -->
   <!-- this option is not documented, but mentioned in release notes and the changelog -->

3. Add an entry to `db/undeploy.json` to undeploy the CAP-generated _.hdbcds_ files:

    ::: code-group
    
    ```json [db/undeploy.json]
    [
      ...,
      "src/gen/**/*.hdbcds"
    ]
    ```
    :::
    

>     <!-- **TODO** Without this entry, during HDI deployment you will get errors like ... -->

Author: We would get errors like this:

```
    Error: "db://E": the object cannot be provided more than once [8212002]
     "src/E.hdbtable": the file would provide it 
     "$cds.merge/E": the deployed file already provides it 
      Merged from "src/E.hdbcds"
```

>     <!-- **TODO** Without this entry, during HDI deployment you will get errors like ... -->

Author: Maybe put this in an info-box? warning-box?

>     <!-- **TODO** Without this entry, during HDI deployment you will get errors like ... -->

Reviewer 2: Have added the error message.
I leave it to @smahati and @renejeglinsky whtherthey want to have it in a box or not ...

>     <!-- **TODO** Without this entry, during HDI deployment you will get errors like ... -->

Reviewer: Added it to a detail box.

add native SQL clauses to the generated .hdbtable files,
or to add "native" HANA CDS clauses to the generated .hdbcds files, respectively.

If you have used these annotations in your model, a simple switchover from hdbcds to hdbtable
very likely is not possible, as such an annotation written for hdbcds in general is not valid
for hdbtable. You have to adapt your model prior to the migration.

As we don't know what clauses you have used, we cannot offer any further guidance here.

## `WITH ASSOCIATIONS`

Associations cause issues in the .hdbcds to .hdbtable handover.
For each entity that has associations, the resulting table or view contains a `WITH ASSOCIATIONS` clause,
representing native HANA associations.

When deploying via hdbcds, associations in a CAP CDS entity are reflected by corresponding associations in
the `hdbcds` file generated by the compiler. Upon deployment, the hdbcds plugin of HDI generates a `CREATE TABLE`
statement, where the associations are represented in a `WITH ASSOCIATIONS` clause.
When deploying via hdbtable, the compiler directly writes the `CREATE TABLE` statements with the `WITH ASSOCIATIONS`
clause into the generated `hdbtable` files.

> These clauses slightly differ, which causes a full table migration when switching from hdbcds to hdbtable.

Reviewer 3: 🥇

---
shorty: Moving from .hdbcds to .hdbtable
synopsis: >

>   This document describes the migration from .hdbcds-based HDI deployment to .hdbtable-based HDI deployment.

Reviewer: AI-generated short description.
```suggestion
The deploy format `hdbcds` for SAP HANA has been deprecated with @sap/cds-compiler@5 and @sap/cds@8, and users are advised to switch to the default format `hdbtable`. This guide provides a step-by-step process for making the switch, including potential issues and workarounds, such as handling annotations `@sql.append` or `sql.prepend`, dealing with associations, multiline doc comments, and temporal data with time slice IDs.
```

# layout: cds-ref
redirect_from: releases/compiler-v2
status: internal
---

# Moving from deploy format `hdbcds` to `hdbtable`

{{ $frontmatter.synopsis }}

::: info

> Not relevant for SAP HANA Cloud.

Author: ```suggestion
If you are using SAP HANA Cloud, you don't have to do anything, as there is no HANA CDS on SAP HANA Cloud.
```

An explaining sentence on why its not relevant maybe?

:::

With @sap/cds-compiler@5 and @sap/cds@8, we have deprecated the deploy format `hdbcds` for SAP HANA,
together with the function [`to.hdbcds`](../node.js/cds-compile#hdbcds).
New CDS features will not be available for deploy format `hdbcds`, and it is going to be removed with one of the
next major releases.

In case your database deployment is still based on `hdbcds`, you should move to the default format `hdbtable`

> with the following 4 steps. This guide assumes you use @sap/cds@7 or higher.

Author: @patricebender if I understood you correctly we need cds@8 to do the no-assoc-step, right?

Reviewer 3: yes because when I tested it with 7.9.2, the `"native_hana_associations": false` was not doing anything

> Info: The deploy format determines only the "medium" how your database model is brought to the database.

Author: TODO: Real info-box?

Reviewer 2: done

If you have used these annotations in your model, a simple switchover from hdbcds to hdbtable
very likely is not possible, as such an annotation written for hdbcds in general is not valid
for hdbtable. You have to adapt your model prior to the migration.

As we don't know what clauses you have used, we cannot offer any further guidance here.

## `WITH ASSOCIATIONS`

Associations cause issues in the .hdbcds to .hdbtable handover.
For each entity that has associations, the resulting table or view contains a `WITH ASSOCIATIONS` clause,
representing native HANA associations.

When deploying via hdbcds, associations in a CAP CDS entity are reflected by corresponding associations in
the `hdbcds` file generated by the compiler. Upon deployment, the hdbcds plugin of HDI generates a `CREATE TABLE`
statement, where the associations are represented in a `WITH ASSOCIATIONS` clause.
When deploying via hdbtable, the compiler directly writes the `CREATE TABLE` statements with the `WITH ASSOCIATIONS`
clause into the generated `hdbtable` and `hdbview` files.
These clauses slightly differ, which causes a full table migration when switching from hdbcds to hdbtable.

> The CAP Java runtime and the CAP Nodejs runtime with the new SAP HANA service (`@cap-js/hana`, default in CDS 8)

Author: ```suggestion
The CAP Java runtime and the CAP Nodejs runtime with the new SAP HANA service (`@cap-js/hana`, default in @sap/cds@8)
```

don't need the `WITH ASSOCIATIONS` clause anymore. This allows us to avoid the full table migration by removing
the `associations` from the `hdbcds` sources __before__ the actual `hdbcds` to `hdbtable` migration.

First switch off the generation of the `associations` (that option accounts for "associations" in the `hdbcds`
sources as well as for the `WITH ASSOCIATIONS` found in the `hdbtable` sources):

::: code-group

```json [.cdsrc.json]
{
  "sql": {
    "native_hana_associations": false
  }
}
```
:::
<!-- this option is available only with CDS 8 -->

Then run a new build and deploy the newly generated `.hdbcds` files.
In contrast to the hdbtable plugin, the hdbcds plugin is able to handle removal of the
native associations without a full table migration.
The resulting database tables and views don't contain any `associations` anymore.

::: warning Requirements

This workaround requires

> * CAP CDS 8,

Author: ```suggestion
* @sap/cds@8,
```

statement, where the doc comments are represented by `COMMENT` clauses.
When deploying via hdbtable, the compiler directly writes the `CREATE TABLE` statements with the `COMMENT`
clauses into the generated `hdbtable` and `hdbview` files.
These `COMMENT` clauses slightly differ, which causes a full table migration when switching from hdbcds to hdbtable.

If you don't actually need the comments in the database, you can remove them as a preparation step
__before__ you do the hdbcds to hdbtable migration.
This is similar to the workaround described above for the `WITH ASSOCIATIONS` clause.

First disable the doc comments by adapting your `.cdsrc.json`:
::: code-group
```json [cdsrc.json]
{
  "hana": {
    "comments": false
  }
}
```
:::

Then run a new build and deploy the newly generated `.hdbcds` files.
The `@Comment` annotations have vanished from the `hdbcds` files, thus
the resulting database tables and views don't contain the `COMMENT` clause anymore.
In contrast to the hdbtable plugin, the hdbcds plugin is able to handle removal of the
`COMMENT`s without a full table migration.

> Note: If you need the comments in the database, this workaround will not help,

Author: TODO: Do as warning block?

Reviewer 2: done

clause into the generated `hdbtable` and `hdbview` files.
These clauses slightly differ, which causes a full table migration when switching from hdbcds to hdbtable.

The CAP Java runtime and the CAP Nodejs runtime with the new SAP HANA service (`@cap-js/hana`, default in CDS 8)
don't need the `WITH ASSOCIATIONS` clause anymore. This allows us to avoid the full table migration by removing
the `associations` from the `hdbcds` sources __before__ the actual `hdbcds` to `hdbtable` migration.

First switch off the generation of the `associations` (that option accounts for "associations" in the `hdbcds`
sources as well as for the `WITH ASSOCIATIONS` found in the `hdbtable` sources):

::: code-group

```json [.cdsrc.json]
{
  "sql": {
    "native_hana_associations": false
  }
}
```
:::
<!-- this option is available only with CDS 8 -->

Then run a new build and deploy the newly generated `.hdbcds` files.
In contrast to the hdbtable plugin, the hdbcds plugin is able to handle removal of the
native associations without a full table migration.
The resulting database tables and views don't contain any `associations` anymore.

> ::: warning Requirements

Reviewer 2: @renejeglinsky I don't get this comment - what is wrong with the original line, it does specify a container type "warning" ?

> ::: warning Requirements

Reviewer: I guess this was a mistake. We can ignore it.
